datadog/resource_datadog_synthetics_test_.go

```diff
@@ -0,0 +1,432 @@
+		},
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice(syntheticsTypes, false),
+			},
+			"request": syntheticsTestRequest(),
+			"request_headers": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+			"assertions": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeMap,
+				},
+			},
+			"devices": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeMap,
+				},
+			},
+			"locations": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"options": syntheticsTestOptions(),
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"message": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			"tags": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"paused": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func syntheticsTestRequest() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeMap,
+		Required: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"method": {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+				"url": {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+				"body": {
+					Type:     schema.TypeString,
+					Optional: true,
+				},
+				"timeout": {
+					Type:     schema.TypeInt,
+					Optional: true,
+					Default:  0,
+				},
+			},
+		},
+	}
+}
+
+func syntheticsTestOptions() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeMap,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"tick_every": {
+					Type:     schema.TypeInt,
+					Required: true,
+				},
+				"min_failure_duration": {
+					Type:     schema.TypeInt,
+					Optional: true,
+				},
+				"min_location_failed": {
+					Type:     schema.TypeInt,
+					Optional: true,
+				},
+			},
+		},
+	}
+}
+
+func resourceDatadogSyntheticsTestCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*datadog.Client)
+
+	syntheticsTest := newSyntheticsTestFromLocalState(d)
+	createdSyntheticsTest, err := client.CreateSyntheticsTest(syntheticsTest)
+	if err != nil {
+		// Note that Id won't be set, so no state will be saved.
+		return fmt.Errorf("error creating synthetics test: %s", err.Error())
+	}
+
+	// If the Create callback returns with or without an error without an ID set using SetId,
+	// the resource is assumed to not be created, and no state is saved.
+	d.SetId(createdSyntheticsTest.GetPublicId())
+
+	// Call resume/pause webservice, because it is a dedicated endpoint apart from classical CRUD operations
+	updateSyntheticsTestLiveness(d, client)
```

Comment:
The `updateSyntheticsTestLiveness` should have a return value for errors. It's ok to return errors from Create/Update functions as mentioned in [1], even when the resource is created/updated. This is called partial state and I think it fits nicely with what is done with the `paused` attribute being set via additional call.

[1] https://www.terraform.io/docs/extend/writing-custom-providers.html#error-handling-amp-partial-state

Reply:
We changed the API to accept directly the status live/paused, so it's not necessary anymore to do an extra call (whose semantics were hard to get right IMO).